libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,890 @@
+          .primary(args_.y_min ? args_.y_min->source : args_.y_max->source)
+          .note("skipping series")
+          .emit(dh);
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Modified from `floor()`
+  auto floor(const series& xs) const -> series {
+    return match(
+      *xs.array,
+      [&](const arrow::DurationArray& array) -> series {
+        auto b
+          = duration_type::make_arrow_builder(arrow::default_memory_pool());
+        check(b->Reserve(array.length()));
+        for (auto i = int64_t{0}; i < array.length(); i++) {
+          if (array.IsNull(i)) {
+            check(b->AppendNull());
+            continue;
+          }
+          const auto val = array.Value(i);
+          const auto count = std::abs(args_.res->inner.count());
+          const auto rem = std::abs(val % count);
+          if (rem == 0) {
+            check(b->Append(val));
+            continue;
+          }
+          const auto floor = val >= 0 ? -rem : rem - count;
+          check(b->Append(val + floor));
+        }
+        return {duration_type{}, finish(*b)};
+      },
+      [&](const arrow::TimestampArray& array) -> series {
+        auto opts = make_round_temporal_options(args_.res->inner);
+        return {time_type{},
+                check(arrow::compute::FloorTemporal(array, std::move(opts)))
+                  .array_as<arrow::TimestampArray>()};
+      },
+      [&](const auto&) -> series {
+        TENZIR_UNREACHABLE();
+      });
+  }
+
+  friend auto inspect(auto& f, chart_operator2& x) -> bool {
+    return f.apply(x.args_);
+  }
+
+private:
+  chart_args args_;
+};
+
+template <chart_type Ty>
+class chart_plugin : public virtual operator_factory_plugin {
+  auto name() const -> std::string override {
+    return fmt::format("chart_{}", to_string(Ty));
+  }
+
+  auto make(invocation inv, session ctx) const
+    -> failure_or<operator_ptr> override {
+    auto args = chart_args{};
+    args.ty = Ty;
+    args.op_loc = inv.self.get_location();
+    if constexpr (Ty == chart_type::bar or Ty == chart_type::pie) {
+      args.limit.inner = 100;
+    }
+    auto y = ast::expression{};
+    auto x_min = std::optional<located<data>>{};
+    auto x_max = std::optional<located<data>>{};
+    auto p = argument_parser2::operator_(name());
+    p.named(Ty == chart_type::pie ? "label" : "x", args.x);
+    p.named(Ty == chart_type::pie ? "value" : "y", y, "any");
+    if constexpr (Ty != chart_type::pie) {
+      p.named("x_min", x_min, "constant");
+      p.named("x_max", x_max, "constant");
+      p.named("y_min", args.y_min);
+      p.named("y_max", args.y_max);
+      p.named("resolution", args.res);
+      p.named("x_log", args.x_log);
+      p.named("y_log", args.y_log);
+    }
+    p.named("group", args.group, "any");
+    if constexpr (Ty == chart_type::area or Ty == chart_type::bar) {
+      p.named_optional("position", args.position);
+    }
+    p.named_optional("_limit", args.limit);
```

Comment:
I think we agreed on adding a check that this does not exceed the default limit, right? I can't find it.

Reply:
It's in the `validate` call.